File: defence_claw/defenseclaw/internal/gateway/connector/platform_support.go
// Copyright 2026 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package connector

import (
	"fmt"
	"runtime"
)

// PlatformSupportStatus is the operator-facing availability state for a
// connector on an operating system.
type PlatformSupportStatus string

const (
	PlatformSupported    PlatformSupportStatus = "supported"
	PlatformPreview      PlatformSupportStatus = "preview"
	PlatformNotCertified PlatformSupportStatus = "not_certified"
	PlatformUnsupported  PlatformSupportStatus = "unsupported"
)

// PlatformSupport includes both the machine-readable state and the reason that
// setup/presentation surfaces show to the operator.
type PlatformSupport struct {
	Status PlatformSupportStatus
	Reason string
}

// proxyConnectors are the chat/LLM-proxy connectors that interpose on model
// traffic through DefenseClaw's local guardrail proxy. Topology and platform
// support are deliberately separate facts: OpenClaw itself has a native
// Windows path, but DefenseClaw does not host its proxy lifecycle on Windows.
var proxyConnectors = map[string]struct{}{
	"openclaw":  {},
	"zeptoclaw": {},
}

// windowsConnectorSupport is the Go source of truth for native Windows
// connector availability. Keep it in exact parity with the Python
// cli/defenseclaw/platform_support.py WINDOWS_CONNECTOR_SUPPORT mapping.
var windowsConnectorSupport = map[string]PlatformSupport{
	"codex": {
		Status: PlatformSupported,
		Reason: "Codex CLI and the DefenseClaw hook entrypoint are certified on native Windows x64.",
	},
	"claudecode": {
		Status: PlatformSupported,
		Reason: "Claude Code with Git for Windows and native hooks is certified on native Windows x64.",
	},
	"cursor": {
		Status: PlatformNotCertified,
		Reason: "The DefenseClaw Cursor integration has not completed native Windows x64 certification.",
	},
	"windsurf": {
		Status: PlatformNotCertified,
		Reason: "The DefenseClaw Windsurf integration has not completed native Windows x64 certification.",
	},
	"geminicli": {
		Status: PlatformNotCertified,
		Reason: "The DefenseClaw Gemini CLI integration has not completed native Windows x64 certification.",
	},
	"copilot": {
		Status: PlatformNotCertified,
		Reason: "The DefenseClaw GitHub Copilot CLI integration has not completed native Windows x64 certification.",
	},
	"antigravity": {
		Status: PlatformNotCertified,
		Reason: "The DefenseClaw Antigravity integration has not completed native Windows x64 certification.",
	},
	"opencode": {
		Status: PlatformNotCertified,
		Reason: "The DefenseClaw OpenCode integration has not completed native Windows x64 certification.",
	},
	"amp": {
		Status: PlatformSupported,
		Reason: "Amp and the DefenseClaw system policy plugin are supported on native Windows x64.",
	},
	"hermes": {
		Status: PlatformNotCertified,
		Reason: "The DefenseClaw Hermes integration has not completed native Windows x64 certification.",
	},
	"openhands": {
		Status: PlatformUnsupported,
		Reason: "OpenHands CLI requires WSL; DefenseClaw does not implement a WSL connector path.",
	},
	"omnigent": {
		Status: PlatformUnsupported,
		Reason: "OmniGent has no supported native Windows terminal/sandbox path for this connector.",
	},
	"openclaw": {
		Status: PlatformUnsupported,
		Reason: "DefenseClaw on Windows is hook-only; OpenClaw integration requires the guardrail proxy.",
	},
	"zeptoclaw": {
		Status: PlatformUnsupported,
		Reason: "ZeptoClaw publishes macOS/Linux builds and its DefenseClaw integration requires the guardrail proxy.",
	},
}

// IsProxyConnector reports whether name is a proxy/chat connector (as opposed
// to a hook-based connector).
func IsProxyConnector(name string) bool {
	_, ok := proxyConnectors[name]
	return ok
}

// ConnectorSupportOnOS returns a supported/preview/not-certified/unsupported
// classification with a human-readable reason. Unknown plugin connectors fail
// closed on Windows pending separate certification.
func ConnectorSupportOnOS(name, goos string) PlatformSupport {
	if goos == "windows" {
		if support, ok := windowsConnectorSupport[name]; ok {
			return support
		}
		return PlatformSupport{
			Status: PlatformNotCertified,
			Reason: "This connector has not completed native Windows x64 certification.",
		}
	}
	return PlatformSupport{
		Status: PlatformSupported,
		Reason: fmt.Sprintf("Connector setup is supported on %s.", goos),
	}
}

// connectorSupportedOnOS reports whether setup/presentation may offer name on
// goos. Preview connectors are deliberately available.
func connectorSupportedOnOS(name, goos string) bool {
	status := ConnectorSupportOnOS(name, goos).Status
	return connectorStatusAvailable(status)
}

func connectorStatusAvailable(status PlatformSupportStatus) bool {
	return status == PlatformSupported || status == PlatformPreview
}

// ConnectorSupportOnHostOS returns the full classification for the host OS.
func ConnectorSupportOnHostOS(name string) PlatformSupport {
	return ConnectorSupportOnOS(name, runtime.GOOS)
}

// ConnectorSupportedOnHostOS reports availability on the current host OS.
func ConnectorSupportedOnHostOS(name string) bool {
	status := ConnectorSupportOnHostOS(name).Status
	return connectorStatusAvailable(status)
}

// CheckPlatformSupport returns the shared operator-facing preview warning or
// unsupported error for a connector on goos. Supported connectors return two
// empty values so callers can preserve their existing control flow.
func CheckPlatformSupport(name, goos string) (string, error) {
	support := ConnectorSupportOnOS(name, goos)
	switch support.Status {
	case PlatformUnsupported:
		return "", fmt.Errorf("connector %q is not supported on %s: %s", name, goos, support.Reason)
	case PlatformNotCertified:
		return "", fmt.Errorf("connector %q is not certified on %s: %s", name, goos, support.Reason)
	case PlatformPreview:
		return fmt.Sprintf("connector %s is preview on %s: %s", name, goos, support.Reason), nil
	default:
		return "", nil
	}
}

// CheckPlatformSupportOnHost applies CheckPlatformSupport to runtime.GOOS.
func CheckPlatformSupportOnHost(name string) (string, error) {
	return CheckPlatformSupport(name, runtime.GOOS)
}

// validateConnectorSupportedOnOS returns the clear setup error used by direct
// connector lifecycle calls. It is injectable by OS for focused tests.
func validateConnectorSupportedOnOS(name, goos string) error {
	_, err := CheckPlatformSupport(name, goos)
	return err
}

func errConnectorUnsupportedOnOS(name, goos string) error {
	return validateConnectorSupportedOnOS(name, goos)
}

File: defence_claw/defenseclaw/cli/defenseclaw/platform_support.py
# Copyright 2026 Cisco Systems, Inc. and its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# SPDX-License-Identifier: Apache-2.0

"""Per-OS connector support — the Python single source of truth.

Windows support is not a boolean derived from connector topology.  A native
agent/runtime and a DefenseClaw integration that can be wired without WSL are
both required. The resulting status is one of ``supported``, ``preview``,
``not_certified``, or ``unsupported`` and always carries a reason.

DefenseClaw runs hook-only on Windows: agents invoke the native Go hook
entrypoint (``defenseclaw-gateway hook``) directly, and there is no Windows
guardrail-proxy lifecycle. The proxy/chat connectors (``openclaw`` and
``zeptoclaw``) therefore cannot run on Windows, so the TUI/CLI must not offer
or accept them there.

This module mirrors ``internal/gateway/connector/platform_support.go``.  Tests
pin the two taxonomies and all Python presentation lists together.  macOS and
Linux retain their historical behavior: every built-in and plugin connector is
offered there.
"""

from __future__ import annotations

import sys
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Literal
from urllib.parse import urlparse

SupportStatus = Literal["supported", "preview", "not_certified", "unsupported"]

SUPPORTED: SupportStatus = "supported"
PREVIEW: SupportStatus = "preview"
NOT_CERTIFIED: SupportStatus = "not_certified"
UNSUPPORTED: SupportStatus = "unsupported"

PROXY_CONNECTORS: frozenset[str] = frozenset({"openclaw", "zeptoclaw"})

LOCAL_OBSERVABILITY_UNSUPPORTED_REASON = "Bundled local observability is unavailable on this operating system."
LOCAL_SPLUNK_UNSUPPORTED_REASON = "Bundled Local Splunk is unavailable on this operating system."
# Compatibility name retained for callers that predate the native controller.
LOCAL_SHELL_STACKS_UNSUPPORTED_REASON = LOCAL_SPLUNK_UNSUPPORTED_REASON


@dataclass(frozen=True)
class ConnectorPlatformSupport:
    """Support classification for one connector on one operating system."""

    status: SupportStatus
    reason: str

    @property
    def available(self) -> bool:
        """Whether setup/presentation may offer this connector."""
        return self.status in {SUPPORTED, PREVIEW}


# Keep in exact parity with the Go ``windowsConnectorSupport`` map. A working
# upstream Windows binary is not sufficient for DefenseClaw certification.
WINDOWS_CONNECTOR_SUPPORT: dict[str, ConnectorPlatformSupport] = {
    "codex": ConnectorPlatformSupport(
        SUPPORTED,
        "Codex CLI and the DefenseClaw hook entrypoint are certified on native Windows x64.",
    ),
    "claudecode": ConnectorPlatformSupport(
        SUPPORTED,
        "Claude Code with Git for Windows and native hooks is certified on native Windows x64.",
    ),
    "cursor": ConnectorPlatformSupport(
        NOT_CERTIFIED,
        "The DefenseClaw Cursor integration has not completed native Windows x64 certification.",
    ),
    "windsurf": ConnectorPlatformSupport(
        NOT_CERTIFIED,
        "The DefenseClaw Windsurf integration has not completed native Windows x64 certification.",
    ),
    "geminicli": ConnectorPlatformSupport(
        NOT_CERTIFIED,
        "The DefenseClaw Gemini CLI integration has not completed native Windows x64 certification.",
    ),
    "copilot": ConnectorPlatformSupport(
        NOT_CERTIFIED,
        "The DefenseClaw GitHub Copilot CLI integration has not completed native Windows x64 certification.",
    ),
    "antigravity": ConnectorPlatformSupport(
        NOT_CERTIFIED,
        "The DefenseClaw Antigravity integration has not completed native Windows x64 certification.",
    ),
    "opencode": ConnectorPlatformSupport(
        NOT_CERTIFIED,
        "The DefenseClaw OpenCode integration has not completed native Windows x64 certification.",
    ),
    "amp": ConnectorPlatformSupport(
        SUPPORTED,
        "Amp and the DefenseClaw system policy plugin are supported on native Windows x64.",
    ),
    "hermes": ConnectorPlatformSupport(
        NOT_CERTIFIED,
        "The DefenseClaw Hermes integration has not completed native Windows x64 certification.",
    ),
    "openhands": ConnectorPlatformSupport(
        UNSUPPORTED,
        "OpenHands CLI requires WSL; DefenseClaw does not implement a WSL connector path.",
    ),
    "omnigent": ConnectorPlatformSupport(
        UNSUPPORTED,
        "OmniGent has no supported native Windows terminal/sandbox path for this connector.",
    ),
    "openclaw": ConnectorPlatformSupport(
        UNSUPPORTED,
        "DefenseClaw on Windows is hook-only; OpenClaw integration requires the guardrail proxy.",
    ),
    "zeptoclaw": ConnectorPlatformSupport(
        UNSUPPORTED,
        "ZeptoClaw publishes macOS/Linux builds and its DefenseClaw integration requires the guardrail proxy.",
    ),
}

WINDOWS_SUPPORTED_CONNECTORS: frozenset[str] = frozenset(
    name for name, support in WINDOWS_CONNECTOR_SUPPORT.items() if support.status == SUPPORTED
)
WINDOWS_PREVIEW_CONNECTORS: frozenset[str] = frozenset(
    name for name, support in WINDOWS_CONNECTOR_SUPPORT.items() if support.status == PREVIEW
)
WINDOWS_NOT_CERTIFIED_CONNECTORS: frozenset[str] = frozenset(
    name for name, support in WINDOWS_CONNECTOR_SUPPORT.items() if support.status == NOT_CERTIFIED
)
WINDOWS_UNSUPPORTED_CONNECTORS: frozenset[str] = frozenset(
    name for name, support in WINDOWS_CONNECTOR_SUPPORT.items() if support.status == UNSUPPORTED
)

WINDOWS_CERTIFIED_ARCHITECTURES: frozenset[str] = frozenset({"amd64"})
WINDOWS_NOT_CERTIFIED_ARCHITECTURES: frozenset[str] = frozenset({"arm64"})
WINDOWS_UNSUPPORTED_FEATURES: frozenset[str] = frozenset(
    {
        "sandbox",
        "enterprise-hooks",
        "openhands",
        "omnigent",
        "openclaw",
        "zeptoclaw",
    }
)


def host_os() -> str:
    """Return a Go-``GOOS``-style token for the current host."""
    return _normalize_os_name(sys.platform)


def _normalize_os_name(os_name: str) -> str:
    value = (os_name or "").strip().lower()
    if value.startswith("win"):
        return "windows"
    if value == "darwin":
        return "darwin"
    if value.startswith("linux"):
        return "linux"
    return value


def is_proxy_connector(name: str) -> bool:
    """Report whether *name* is a proxy/chat connector."""
    return name in PROXY_CONNECTORS


def connector_platform_support(
    name: str,
    os_name: str | None = None,
) -> ConnectorPlatformSupport:
    """Return the status and reason for *name* on *os_name*.

    Unknown/plugin connectors require separate native Windows certification.
    macOS and Linux preserve their historical supported behavior.
    """
    resolved_os = host_os() if os_name is None else _normalize_os_name(os_name)
    if resolved_os == "windows":
        return WINDOWS_CONNECTOR_SUPPORT.get(
            name,
            ConnectorPlatformSupport(
                NOT_CERTIFIED,
                "This connector has not completed native Windows x64 certification.",
            ),
        )
    return ConnectorPlatformSupport(
        SUPPORTED,
        f"Connector setup is supported on {resolved_os or 'this platform'}.",
    )


def connector_support_status(name: str, os_name: str | None = None) -> SupportStatus:
    """Return only the support status for presentation/serialization."""
    return connector_platform_support(name, os_name).status


def connector_support_reason(name: str, os_name: str | None = None) -> str:
    """Return the operator-facing reason for the connector's status."""
    return connector_platform_support(name, os_name).reason


def connector_supported_on_os(name: str, os_name: str | None = None) -> bool:
    """Report whether *name* may be offered/used on *os_name*.

    Preview connectors are deliberately available. Not-certified and
    unsupported connectors are hidden from pickers and rejected by setup.
    """
    return connector_platform_support(name, os_name).available


def connector_preview_on_os(name: str, os_name: str | None = None) -> bool:
    """Report whether *name* is available as a preview on *os_name*."""
    return connector_support_status(name, os_name) == PREVIEW


def local_observability_stack_supported(os_name: str | None = None) -> bool:
    """Whether the shared Python-backed observability controller is available."""

    resolved_os = host_os() if os_name is None else _normalize_os_name(os_name)
    return resolved_os in {"windows", "darwin", "linux"}


def local_splunk_stack_supported(os_name: str | None = None) -> bool:
    """Whether bundled Local Splunk has a certified lifecycle controller."""

    resolved_os = host_os() if os_name is None else _normalize_os_name(os_name)
    return resolved_os in {"windows", "darwin", "linux"}


def local_shell_stacks_supported(os_name: str | None = None) -> bool:
    """Backward-compatible alias for bundled Local Splunk availability."""

    return local_splunk_stack_supported(os_name)


def is_local_observability_stack_destination(
    *,
    name: str = "",
    preset_id: str = "",
    kind: str = "",
    endpoint: str = "",
) -> bool:
    """Classify config/runtime state owned by bundled local observability."""

    if preset_id == "local-otlp" or name in {"local-observability", "local-otlp-logs"}:
        return True
    return False


def is_local_splunk_stack_destination(
    *,
    preset_id: str = "",
    kind: str = "",
    endpoint: str = "",
) -> bool:
    """Classify loopback Splunk HEC state owned by the local Splunk stack."""

    if kind != "splunk_hec":
        return False
    if preset_id == "splunk-enterprise":
        return False
    parsed = urlparse(endpoint if "://" in endpoint else f"//{endpoint}")
    return (parsed.hostname or "").lower() in {"localhost", "127.0.0.1", "::1"}


def is_local_shell_stack_destination(
    *,
    name: str = "",
    preset_id: str = "",
    kind: str = "",
    endpoint: str = "",
) -> bool:
    """Backward-compatible union classifier for older callers."""

    return is_local_observability_stack_destination(
        name=name, preset_id=preset_id, kind=kind, endpoint=endpoint
    ) or is_local_splunk_stack_destination(
        preset_id=preset_id,
        kind=kind,
        endpoint=endpoint,
    )


def destination_platform_unsupported(
    *,
    name: str = "",
    preset_id: str = "",
    kind: str = "",
    endpoint: str = "",
    os_name: str | None = None,
) -> bool:
    """Whether a destination belongs to a local stack unavailable here."""

    return (
        not local_observability_stack_supported(os_name)
        and is_local_observability_stack_destination(
            name=name,
            preset_id=preset_id,
            kind=kind,
            endpoint=endpoint,
        )
    ) or (
        not local_splunk_stack_supported(os_name)
        and is_local_splunk_stack_destination(
            preset_id=preset_id,
            kind=kind,
            endpoint=endpoint,
        )
    )


def supported_connectors(names: Iterable[str], os_name: str | None = None) -> list[str]:
    """Filter *names* to supported/preview entries, preserving order."""
    return [n for n in names if connector_supported_on_os(n, os_name)]
